lib/spack/spack/fetch_strategy.py

```diff
@@ -740,11 +745,429 @@ def __str__(self):
+                raise InvalidGitFetchStageConfig(
+                    'argument submodules_delete={0!r} must be a list of str'
+                    .format(submodules_delete))
+            get_full_repo = cls._extract_bool(kwargs, 'get_full_repo', False)
+        except InvalidGitFetchStageConfig as e:
+            raise InvalidGitFetchStageConfig(
+                'failed to parse {0} from kwargs {1}: {2}'
+                .format(cls.__name__, kwargs, e))
+        return cls(submodules=submodules,
+                   submodules_delete=submodules_delete,
+                   get_full_repo=get_full_repo)
+
+
+class ConfiguredGit(object):
+    """Caches the version for the `git` executable, and adds compatibility args.
+
+    Use the `.from_executable()` method to perform this checking:
+
+        from spack.fetch_strategy import ConfiguredGit
+        from spack.util.Executable import which
+
+        git = ConfiguredGit.from_executable(which('git', required=True))
+        assert git.version == '2.31.1'
+
+    And note that the `__call__` method delegates to the inner executable from `which`:
+
+        assert git('--version', output=str) == 'git version 2.31.1'
+    """
+    git = None                 # type: Executable
+    version = None             # type: Version
+
+    def __init__(self, git, version):
+        # type: (Executable, Version) -> None
+        self.git = git
+        self.version = version
+
+    def __repr__(self):
+        return '{0}(git={1!r}, version={2!r})'.format(
+            type(self).__name__, self.git, self.version)
+
+    def __call__(self, *args, **kwargs):
+        return self.git(*args, **kwargs)
+
+    @staticmethod
+    def _get_git_version(git):
+        # type: (Executable) -> Version
+        vstring = git('--version', output=str).lstrip('git version ')
+        return Version(vstring)
+
+    @classmethod
+    @memoized
+    def from_executable(cls, git):
+        # type: (Executable) -> ConfiguredGit
+        """Extract the version from a git executable and add defaults for compatibility.
+
+        This method being `@memoized` means that with the current implementation of
+        `Executable.__eq__`, any two `Executable` instances pointing to the same
+        executable file path will return a cached result from this method, which will
+        avoid traversing the filesystem.
+        """
+        version = cls._get_git_version(git)
+
+        # Disable advice for a quieter fetch
+        # https://github.com/git/git/blob/master/Documentation/RelNotes/1.7.2.txt
+        if version >= Version('1.7.2'):
+            git.add_default_arg('-c')
+            git.add_default_arg('advice.detachedHead=false')
+
+        # If the user asked for insecure fetching, make that work
+        # with git as well.
+        if not spack.config.get('config:verify_ssl'):
+            git.add_default_env('GIT_SSL_NO_VERIFY', 'true')
+
+        return cls(git, version)
```

Comment:
This logic isn't working properly on macos.

```
$ spack -d fetch hwloc@master
<snip>
==> [2021-06-10-18:36:23.003847] '/usr/bin/git' '--version'
Traceback (most recent call last):
  File "/Users/becker33/spack/bin/spack", line 77, in <module>
    sys.exit(spack.main.main())
  File "/Users/becker33/spack/lib/spack/spack/main.py", line 771, in main
    return _invoke_command(command, parser, args, unknown)
  File "/Users/becker33/spack/lib/spack/spack/main.py", line 499, in _invoke_command
    return_val = command(parser, args)
  File "/Users/becker33/spack/lib/spack/spack/cmd/fetch.py", line 86, in fetch
    package.do_fetch()
  File "/Users/becker33/spack/lib/spack/spack/package.py", line 1372, in do_fetch
    self.stage.fetch(mirror_only, err_msg=err_msg)
  File "/Users/becker33/spack/lib/spack/spack/util/pattern.py", line 24, in __call__
    for item in self.container]
  File "/Users/becker33/spack/lib/spack/spack/util/pattern.py", line 24, in <listcomp>
    for item in self.container]
  File "/Users/becker33/spack/lib/spack/spack/stage.py", line 490, in fetch
    self.fetcher.fetch()
  File "/Users/becker33/spack/lib/spack/spack/fetch_strategy.py", line 78, in wrapper
    return fun(self, *args, **kwargs)
  File "/Users/becker33/spack/lib/spack/spack/fetch_strategy.py", line 1335, in fetch
    current_expanded_ref = self._ensure_local_ref()
  File "/Users/becker33/spack/lib/spack/spack/fetch_strategy.py", line 1278, in _ensure_local_ref
    current_expanded_ref = self._maybe_expand_ref()
  File "/Users/becker33/spack/lib/spack/spack/fetch_strategy.py", line 1259, in _maybe_expand_ref
    return self.canonical_git_repo.expand_commit_hash(self.ref)
  File "/Users/becker33/spack/lib/spack/llnl/util/lang.py", line 207, in _memoized_function
    func.cache[args] = func(*args)
  File "/Users/becker33/spack/lib/spack/spack/fetch_strategy.py", line 1226, in canonical_git_repo
    configured_git = ConfiguredGit.from_executable(git_exe)
  File "/Users/becker33/spack/lib/spack/llnl/util/lang.py", line 207, in _memoized_function
    func.cache[args] = func(*args)
  File "/Users/becker33/spack/lib/spack/spack/fetch_strategy.py", line 980, in from_executable
    version = cls._get_git_version(git)
  File "/Users/becker33/spack/lib/spack/spack/fetch_strategy.py", line 967, in _get_git_version
    return Version(vstring)
  File "/Users/becker33/spack/lib/spack/spack/version.py", line 165, in __init__
    raise ValueError("Bad characters in version string: %s" % string)
ValueError: Bad characters in version string: 2.21.0 (Apple Git-122.2)

Reply:
I added the method `_strip_apple_git_suffix` in a062080f75111be9693f264b2ad5f9f715ea1221 and this allows the command `spack fetch -d hwloc@master` (or indeed just `spack install hwloc@master`) to work correctly on macOS.